subsys/pm/device_runtime.c

```diff
@@ -127,29 +122,48 @@ static int pm_device_request(const struct device *dev,
 						  PM_DEVICE_STATE_SUSPEND,
 						  NULL, NULL);
 		}
-		ret = 0;
 		goto out;
 	}
 
+	(void)k_mutex_lock(&dev->pm->lock, K_FOREVER);
+
+	if (!dev->pm->enable) {
+		ret = -ENOTSUP;
+		goto out_unlock;
+	}
+
+	if (target_state == PM_DEVICE_STATE_ACTIVE) {
+		dev->pm->usage++;
+		if (dev->pm->usage < 0) {
+			goto out_unlock;
+		}
+	} else {
+		dev->pm->usage--;
+		if (dev->pm->usage > 1) {
+			goto out_unlock;
+		}
+	}
+
 	(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
```

Comment:
Coming back to this, synchronous calls should not go through the work queue, this was a major flaw in the current implementation that we wanted to change.

Reply:
The synchronous case can just wait the callback indeed. Let me just confirm.